std_execution.bs

```diff
@@ -553,6 +553,293 @@ merged into a single one in an asynchronous fashion.
 `transfer(when_all(inputs...), scheduler)`, but one that allows for better efficiency through customization.
 
+## Senders support cancellation ## {#design-cancellation}
+
+Senders are often used in scenarios where the application may be concurrently executing
+multiple strategies for achieving some program goal. When one of these strategies succeeds
+(or fails) it may not make sense to continue pursuing the other strategies as their results
+are no longer useful.
+
+For example, we may want to try to simultaneously connect to multiple network servers and use
+whichever server responds first. Once the first server responds we no longer need to continue
+trying to connect to the other servers.
+
+Ideally, in these scenarios, we would somehow be able to request that those other strategies
+stop executing promptly so that their resources (e.g. cpu, memory, I/O bandwidth) can be
+released and used for other work.
+
+While the design of senders has support for cancelling an operation before it starts
+by simply destroying the sender or the operation-state returned from `execution::connect()`
+before calling `execution::start()`, there also needs to be a standard, generic mechanism
+to ask for an already-started operation to complete early.
+
+The ability to be able to cancel in-flight operations is fundamental to supporting some kinds
+of generic concurrency algorithms.
+
+For example:
+* a generic `timeout(src, duration)` algorithm needs to be able to cancel the `src` operation after the timeout duration has elapsed.
+* a `first_successful(ops...)` algorithm should cancel the other operations as soon as one operation completes successfuly
+* a `when_all(ops...)` algorithm should cancel other operations as soon as one operation fails
+* a `stop_when(src, trigger)` algorithm should cancel `src` if `trigger` completes first and cancel `trigger` if `src` completes first
+
+
+The mechanism used for communcating cancellation-requests, or stop-requests, needs to have a uniform interface
+so that generic algorithms that compose sender-based operations, such as the ones listed above, are able to
+communicate these cancellation requests to senders that they don't know anything about.
+
+The design is intended to be composable so that cancellation of higher-level operations can propagate
+those cancellation requests through intermediate layers to lower-level operations that need to actually
+respond to the cancellation requests.
+
+For example, we can compose the algorithms mentioned above so that child operations
+are cancelled when any one of the multiple cancellation conditions occurs:
+<pre highlight=c++>
+sender auto composed_cancellation_example(auto query) {
+  return stop_when(
+    timeout(
+      when_all(
+        first_successful(
+          query_server_a(query),
+          query_server_b(query)),
+        load_file("some_file.jpg")),
+      5s),
+    cancelButton.on_click());
+}
+</pre>
```

Comment:
This example creates a precedent for the rest of the paper: I believe that other than the lines that obtain schedulers from some handwaved sources, the rest of the examples in this paper only actually use other facilities in this paper. The only one that does something else is the `async_read` example, but it treats it as a basis operation of sorts, presumably from another library.

This, however, uses at least three things that certainly should be a part of *this* library if we standardize them: `stop_when`, `timeout`, and `first_successful`. I'm somewhat worried about this; I believe that this makes it no longer possible for us to claim that everything in this paper is self-contained, because now we have an example that clearly requires additional sender machinery to even explain what it tries to achieve.

Reply:
Yes, I think for cancellation to be usable and useful in general, that algorithms like `timeout()`, `first_successful` and `stop_when` should be included in this paper. Although for `timeout()`, this also implies the need for a `time_scheduler` concept that provides the ability to create time-based schedule-senders.

It's not particularly compelling for including cancellation support if the only operation that makes use of cancellation is `when_all`.

Reply:
I'm not sure `stop_when` adds to the example. `when_all` and `timeout` seem sufficient